### PR TITLE
[Bug] Exclude Snapshot from adjoint backwards pass

### PR DESCRIPTION
**Pull request by dime10**
The adjoint diff-method on `QubitDevice`s bypasses the normal `dev.apply` function and instead invokes `dev._apply_operation` for all operations on the tape. This leads to a bug with the `Snapshot` operation as implemented on the `default.qubit` device.
See the following example:

```py
dev = qml.device("default.qubit", wires=1)

@qml.qnode(dev, diff_method="adjoint")
def circuit():
    qml.Hadamard(0)
    qml.Snapshot()
    return qml.expval(qml.PauliZ(0))

circuit()
```
```
ValueError: cannot reshape array of size 4 into shape ()
```

### CODE REVIEW

File: pennylane/_qubit_device.py [hunk 1]
```diff
@@ -942,7 +942,7 @@ def adjoint_jacobian(self, tape, starting_state=None, use_device_state=False):
                         'the "adjoint" differentiation method'
                     )
             else:
-                if op.name not in ("QubitStateVector", "BasisState"):
+                if op.name not in ("QubitStateVector", "BasisState", "Snapshot"):
```
Comment (antalszava) [inline pennylane/_qubit_device.py:945]: Out of scope, but wondering: would this be an issue with Barrier too? :thinking:

**Review by antalszava — APPROVED** [review]
Looks good, thanks for the quick fix! :tada:

File: pennylane/_qubit_device.py (hunk 1 above)
Comment (dime10) [reply pennylane/_qubit_device.py:945]: Hmm good question, I'm not sure about other (hardware?) qubit devices, but for our simulator Barrier is an unsupported operation so it is expanded away.

**Review by mlxd — APPROVED** [review]
Nice one. Thanks @dime10

**Comment by dime10** [comment]
@antalszava CodeFactor seems hung up. I ran pylint locally, no new issues.

**Pull request merged**